File: CSharp/Yow.CoD.Finance.Domain/Model/AggregateRoot.cs
using System;
using System.Collections.Generic;
using Yow.CoD.Finance.Domain.Contracts;

namespace Yow.CoD.Finance.Domain.Model
{
    public abstract class AggregateRoot
    {
        private readonly List<Event> _uncommittedEvents = new List<Event>();
        private readonly Dictionary<Type, Action<object>> _handlers =new Dictionary<Type, Action<object>>();

        protected AggregateRoot(Guid id)
        {
            Id = id;
        }

        public Guid Id { get; }
        public int Version { get; private set; }

        /// <summary>
        /// Apply events to rehydrate the instance from existing events
        /// </summary>
        /// <param name="payload">An event to apply to the instance.</param>
        public void ApplyEvent(Event payload)
        {
            _handlers[payload.GetType()](payload);
            Version++;
        }

        /// <summary>
        /// Get all events that have been emmitted but not yet saved.
        /// </summary>
        /// <returns>An array of uncommitted events.</returns>
        public Event[] GetUncommittedEvents()
        {
            return _uncommittedEvents.ToArray();
        }

        /// <summary>
        /// Acknowledges the uncommitted events as committed.
        /// </summary>
        public void ClearUncommittedEvents()
        {
            _uncommittedEvents.Clear();
        }

        protected void AddEvent(Event uncommittedEvent)
        {
            _uncommittedEvents.Add(uncommittedEvent);
            ApplyEvent(uncommittedEvent);
        }

        protected void RegisterHandler<T>(Action<T> handler)
        {
            _handlers.Add(typeof(T), e=>handler((T)e));
        }
    }
}

File: java/src/main/java/com/leecampbell/cod/domain/model/AggregateRoot.java
package com.leecampbell.cod.domain.model;

import com.leecampbell.cod.domain.contracts.DomainEvent;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.util.*;

public abstract class AggregateRoot {

    private final UUID id;
    private final List<DomainEvent> uncommittedEvents = new ArrayList<DomainEvent>();
    private int version;

    protected AggregateRoot(UUID id) {
        this.id = id;
    }

    public UUID id(){
        return id;
    }

    public int version() {
        return version;
    }

    public void ApplyEvent(DomainEvent payload) {
        try {
            Method handler = getMethod(payload);
            handler.setAccessible(true);
			handler.invoke(this, payload);
		} catch (IllegalAccessException e) {
			e.printStackTrace();
		} catch (InvocationTargetException e) {
			e.printStackTrace();
		} catch (NoSuchMethodException e) {
			e.printStackTrace();
        }
        
        version++;
    }

    public DomainEvent[] GetUncommittedEvents() {
        DomainEvent[] snapshot = new DomainEvent[uncommittedEvents.size()];
        snapshot = uncommittedEvents.toArray(snapshot);
        return snapshot;
    }

    public void ClearUncommittedEvents() {
        uncommittedEvents.clear();
    }
    
    protected void AddEvent(DomainEvent uncommittedEvent) {
        uncommittedEvents.add(uncommittedEvent);
        ApplyEvent(uncommittedEvent);
    }

    private Method getMethod(DomainEvent payload) throws NoSuchMethodException, SecurityException {
        Class<? extends AggregateRoot> rootType = this.getClass();
        Class<? extends DomainEvent> eventType = payload.getClass();

        return rootType.getDeclaredMethod("handle", eventType);
    }
}